pkg/sdkserver/localsdk_test.go

```diff
@@ -607,6 +610,130 @@ func TestLocalSDKServerPlayerConnectAndDisconnect(t *testing.T) {
+	count := int64(1)
+	capacity := int64(100)
+	counters := map[string]agonesv1.CounterStatus{
+		name: {Count: count, Capacity: capacity},
+	}
+	fixture := &agonesv1.GameServer{
+		ObjectMeta: metav1.ObjectMeta{Name: "stuff"},
+		Status: agonesv1.GameServerStatus{
+			Counters: counters,
+		},
+	}
+
+	path, err := gsToTmpFile(fixture)
+	assert.NoError(t, err)
+	l, err := NewLocalSDKServer(path, "")
+	assert.Nil(t, err)
+
+	stream := newGameServerMockStream()
+	go func() {
+		err := l.WatchGameServer(&sdk.Empty{}, stream)
+		assert.Nil(t, err)
+	}()
+	assertInitialWatchUpdate(t, stream)
+
+	// wait for watching to begin
+	err = wait.Poll(time.Second, 10*time.Second, func() (bool, error) {
+		found := false
+		l.updateObservers.Range(func(_, _ interface{}) bool {
+			found = true
+			return false
+		})
+		return found, nil
+	})
+	assert.NoError(t, err)
+
+	counterRequest := alpha.GetCounterRequest{Name: name}
+	expected := &alpha.Counter{Name: name, Count: &count, Capacity: &capacity}
+	counter, err := l.GetCounter(context.Background(), &counterRequest)
+	assert.NoError(t, err)
+	if diff := cmp.Diff(expected, counter, protocmp.Transform()); diff != "" {
+		t.Errorf("unexpected difference:\n%v", diff)
+	}
+}
+
+func TestLocalSDKServerUpdateCounter(t *testing.T) {
+	t.Parallel()
+
+	runtime.FeatureTestMutex.Lock()
+	defer runtime.FeatureTestMutex.Unlock()
+	assert.NoError(t, runtime.ParseFeatures(string(runtime.FeatureCountsAndLists)+"=true"))
+
+	name := "sessions"
+	counters := map[string]agonesv1.CounterStatus{
+		name: {Count: 1, Capacity: 100},
+	}
+	fixture := &agonesv1.GameServer{
+		ObjectMeta: metav1.ObjectMeta{Name: "stuff"},
+		Status: agonesv1.GameServerStatus{
+			Counters: counters,
+		},
+	}
+
+	path, err := gsToTmpFile(fixture)
+	assert.NoError(t, err)
+	l, err := NewLocalSDKServer(path, "")
+	assert.Nil(t, err)
+
+	stream := newGameServerMockStream()
+	go func() {
+		err := l.WatchGameServer(&sdk.Empty{}, stream)
+		assert.Nil(t, err)
+	}()
+	assertInitialWatchUpdate(t, stream)
+
+	// wait for watching to begin
+	err = wait.Poll(time.Second, 10*time.Second, func() (bool, error) {
+		found := false
+		l.updateObservers.Range(func(_, _ interface{}) bool {
+			found = true
+			return false
+		})
+		return found, nil
+	})
+	assert.NoError(t, err)
+
+	ONE := int64(1)
+	HUNDRED := int64(100)
+	notupdated := &alpha.Counter{Name: name, Count: &ONE, Capacity: &HUNDRED}
+	startingCounter, err := l.GetCounter(context.Background(), &alpha.GetCounterRequest{Name: "sessions"})
+	assert.NoError(t, err)
+	if diff := cmp.Diff(notupdated, startingCounter, protocmp.Transform()); diff != "" {
+		t.Errorf("unexpected difference:\n%v", diff)
+	}
+
+	count := int64(9)
+	capacity := int64(999)
+	// Note: the field mask filter in UpdateCounter will remove fields from the request object that
+	// are not included in the FieldMask Paths. Here `Name` and `Capacity` will be removed from
+	// the `updated` object after we call `l.UpdateCounter()`.
```

Comment:
What am I missing? I was expecting to see an `expected` that only had `count`, but I see all fields present on line 722, not filtered down?

Reply:
If you were to print out the `updated` object that was originally passed into the UpdateCounterRequest there would only be the `count` field as that particular object would have been filtered down during the UpdateCounter. 

However the UpdateCounter returns the entire Counter object as it was written back to the gameserver, which is why `expected` has all fields, showing the original `name` and `capacity` values and modified `count` value. Is there a clearer way of writing that comment?

Reply:
Ah, now I think I understand. But if `UpdateCounter` is destructive to `updated`, that's probably a bug. In general I expect RPC calls not to destroy the things I'm passing. Maybe fix that and add a test as well?

I would expect to see `expected` (well, `got`) different than `updated` (which should probably be called `update`, the past tense is odd) given that the field mask doesn't have all the fields.

Reply:
It might be, it might not be. Maybe they send the update then cache the updated value? It's an SDK, it just seems non-idiomatic to muck with the caller's structure in this case. (We don't have `const` but this would be `const` in other languages, I feel like.)

Reply:
I realized that on my previous review, I forgot I was reading the server-side code, and nothing the caller sends would change. Sorry, you're right, it's ok this is destroyed.